### ISSUE TITLE
TelePistons adds `telepistons:piston_arm` to the block registry, making it no longer a pure client-side mod.

### ISSUE THREAD

**Issue by akemin-dayo** [Tectato/Telepistons#7]
TelePistons adds `telepistons:piston_arm` to the block registry, making it no longer a pure client-side mod.

(I actually never knew this was how you were modifying the model, as I had always thought you were just changing the appearance of `minecraft:piston_head`.)

While this _generally_ doesn't cause issues for most people's use-cases (the overwhelming majority play on multiplayer servers, or the local singleplayer server), this design decision _does_ have some negative effects on a rarely-used feature of Minecraft — the "Open to LAN" feature of the local singleplayer server.

To be fair, even _I_ don't use this feature (or the local singleplayer server in general, as I always play on a server I run). The only reason why I even noticed this issue was because I was curious if [ViaFabric](https://github.com/ViaVersion/ViaFabric)'s server-side protocol translation would be applied to the "Open to LAN" feature (it is), or if it only works with dedicated Fabric/Quilt servers.

I honestly wasn't sure if I should even submit this issue in the first place, since I… _think_ it'd require a full redesign of how TelePistons works, but I figured I might as well still try and bring this to your attention. (I'm not too experienced with Minecraft client modifications — I'm most familiar with Bukkit server-side plugins ;P)

---

Here's a list of issues this causes:

* Vanilla or vanilla-compliant (Quilt, Fabric, Forge) Minecraft clients that do _not_ have TelePistons installed will either crash, or otherwise be unable to join an "Open to LAN" server hosted by another client that _does_ have TelePistons installed as long as an activated piston (and/or a `minecraft:piston_head`) is in the world.
* The presence of `telepistons:piston_arm` in the registry results in severely broken behaviour from the ViaFabric (ViaVersion) server-side component. When a newer Minecraft version joins the server, all the blocks are mapped incorrectly. It also breaks ViaBackwards entirely — older Minecraft versions are completely unable to join since registry remapping completely fails.

Again, it's a pretty obscure Minecraft feature, but at the very least perhaps this is something that should be mentioned somewhere so people are aware of it…?

**Comment by Tectato** [comment]
Interesting, I hadn't considered the implications for LAN worlds or ViaVersion of adding that block. It's only used for clientside rendering during the moving animation and never actually put in the world so I don't think it would crash someone without the mod, but it may still make the local server not vanilla-compliant or something. This was my first mod so I only used the rendering methods at my immediate disposal which required a `BlockState` as an input, i.e. a registered block to render, so I had to add it to the registry. I'll look into other options for rendering the model as-is without having to make it it's own block, this was on my To-Do list anyways

**Comment by akemin-dayo** [comment]
> It's only used for clientside rendering during the moving animation and never actually put in the world so I don't think it would crash someone without the mod

I've definitely had both a Fabric client and a vanilla client crash while messing with pistons (both the full block and also just the `minecraft:piston_head`) trying to write up this issue report, though it doesn't really reproduce reliably for whatever strange reason:
```
Error executing task on Client
java.lang.IllegalArgumentException: No value with id 20348
	at Not Enough Crashes deobfuscated stack trace.(1.18.2+build.4)
	at net.minecraft.util.collection.IndexedIterable.getOrThrow(IndexedIterable:19)
	at net.minecraft.world.chunk.BiMapPalette.readPacket(BiMapPalette:74)
	at net.minecraft.world.chunk.PalettedContainer.readPacket(PalettedContainer:162)
	at net.minecraft.world.chunk.ChunkSection.fromPacket(ChunkSection:205)
	at net.minecraft.world.chunk.WorldChunk.loadFromPacket(WorldChunk:445)
	at net.minecraft.client.world.ClientChunkManager.loadChunkFromPacket(ClientChunkManager:126)
	at net.minecraft.client.network.ClientPlayNetworkHandler.loadChunk(ClientPlayNetworkHandler:661)
	at net.minecraft.client.network.ClientPlayNetworkHandler.onChunkData(ClientPlayNetworkHandler:656)
	at net.minecraft.network.packet.s2c.play.ChunkDataS2CPacket.apply(ChunkDataS2CPacket:45)
	at net.minecraft.network.packet.s2c.play.ChunkDataS2CPacket.apply(ChunkDataS2CPacket:12)
	at net.minecraft.network.NetworkThreadUtils.method_11072(NetworkThreadUtils:40)
```

> I'll look into other options for rendering the model as-is without having to make it it's own block, this was on my To-Do list anyways

Sounds good! Would be nice if such a change were able to be backported to 1.18.x/1.17.x/1.16.5 too, but I understand if that's a lot of work.

**Comment by Tectato** [comment]
Alright the necessary changes have been made and i've backported the new version to 1.18.2 and 1.16.5. I reckon almost everyone playing on 1.17 has made the switch to 1.18 by now anyways. Let me know if there are still any issues with this

**Comment by akemin-dayo** [comment]
I do think it's worth making a backport to 1.17.x simply because that version is notable for being the last version of Minecraft with classic ore distribution / world height, so a lot of people gravitate towards that version for that reason.

Anyway, I'll test the new changes in a bit!

**Comment by Tectato** [comment]
Ok i did the 1.17.1 backport as well, just gotta wait for the files to get approved on CurseForge and Modrinth now...

**Comment by akemin-dayo** [comment]
I haven't gotten around to testing the "Open to LAN" game server yet, but here's some things I've noticed with this update:

* Your 1.16.x jar is compiled with the incorrect JDK (JDK 17, aka class version 61), making it impossible to load on that version with the default JRE 8 (which most people will be using) — make sure your 1.17.x jar is compiled with JDK 16 (class version 60) while you're at it, too.
* CurseForge (and by extension, I guess Modrinth?) metadata should be modified to mark 1.19.2, 1.19.1, 1.19, 1.18.2, and 1.18.1 as compatible (those are just the versions I've tested).

**Comment by Tectato** [comment]
1.19 is the main branch and i pushed 1.17 just now. completely forgot about the JDK thing, quite inconvenient

**Comment by akemin-dayo** [comment]
Ah oops. I'm dumb and missed the fact that the main branch was 1.19.x. l o l

Anyway, I've submitted PRs #9 and #8 that fix support for 1.17.x and 1.16.x respectively. See the relevant PRs for more information (1.16.x necessitated a minor code change due to language featureset change.)

You shouldn't need to change your locally installed JDK version to build these.
